packages/alfa-media/src/media.ts

```diff
@@ -312,15 +675,238 @@ export namespace Media {
+            if (value.compare(minimum.value) <= 0) {
+              return false;
+            }
+          } else {
+            if (value.compare(minimum.value) < 0) {
+              return false;
+            }
+          }
+        }
+
+        for (const maximum of this._maximum) {
+          if (maximum.isInclusive) {
+            if (value.compare(maximum.value) >= 0) {
+              return false;
+            }
+          } else {
+            if (value.compare(maximum.value) > 0) {
+              return false;
+            }
+          }
+        }
+
+        return true;
+      }
+
+      public hasValue<U extends T>(
+        refinement: Refinement<T, U>
+      ): this is Discrete<U> {
+        return (
+          this._minimum.every((bound) => refinement(bound.value)) &&
+          this._maximum.every((bound) => refinement(bound.value))
+        );
+      }
+
+      public toJSON(): Range.JSON<T> {
+        return {
+          type: "range",
+          minimum: this._minimum.map((bound) => bound.toJSON()).getOr(null),
+          maximum: this._maximum.map((bound) => bound.toJSON()).getOr(null),
+        };
+      }
+    }
+
+    export namespace Range {
+      export interface JSON<T> {
+        [key: string]: json.JSON;
+        type: "range";
+        minimum: Bound.JSON<T> | null;
+        maximum: Bound.JSON<T> | null;
+      }
+
+      export function isRange<T>(value: unknown): value is Range<T> {
+        return value instanceof Range;
+      }
+    }
+
+    export const {
+      of: range,
+      minimum: minimumRange,
+      maximum: maximumRange,
+      isRange,
+    } = Range;
+
+    export class Bound<T = unknown>
+      implements Functor<T>, Serializable<Bound.JSON<T>> {
+      public static of<T>(value: T, isInclusive: boolean): Bound<T> {
+        return new Bound(value, isInclusive);
+      }
+
+      private readonly _value: T;
+      private readonly _isInclusive: boolean;
+
+      private constructor(value: T, isInclusive: boolean) {
+        this._value = value;
+        this._isInclusive = isInclusive;
+      }
+
+      public get value(): T {
+        return this._value;
+      }
+
+      public get isInclusive(): boolean {
+        return this._isInclusive;
+      }
+
+      public map<U>(mapper: Mapper<T, U>): Bound<U> {
+        return new Bound(mapper(this._value), this._isInclusive);
+      }
+
+      public toJSON(): Bound.JSON<T> {
+        return {
+          value: Serializable.toJSON(this._value),
+          isInclusive: this._isInclusive,
+        };
+      }
+    }
+
+    export namespace Bound {
+      export interface JSON<T> {
+        [key: string]: json.JSON;
+        value: Serializable.ToJSON<T>;
+        isInclusive: boolean;
+      }
+    }
+
+    export const { of: bound } = Bound;
+  }
+
   /**
    * @see https://drafts.csswg.org/mediaqueries/#media-condition
    */
-  export class Condition implements Equatable, Serializable {
+  export class Condition
```

Comment:
I'm not too keen on calling this class `Condition` given that it does not corresponds to what CSS calls "media condition" nor to how we use it.
The link just before the class actually points to a list whose first item is `Negation` (hence, a CSS media condition, but not a `Condition`), so not really explaining what the class does.
Similarly, `Negation` has a `condition` field which is not a `Condition` but a `Condition | Feature | Negation` hinting that the name is not good.

Reply:
A "media condition" is any combination of features, negations, and other conditions. Yes, `Negation` should likely fall under "media condition", but the point is that a "media condition" is a _combination_ of multiple "media features", as per the prose in the specification.

I'd be happy to use the term "media expression" to cover over "media condition, feature, and negation" though, which would also mean that the `Negation#condition` would become `Negation#expression` but "media condition" should remain the combination of _multiple_ "media expressions".

Reply:
I think I found a better approach that aligns more closely with the specification prose: `Condition` is now `And | Or | Not`, which matches the description of conditions being features combined with boolean algebra.